### ISSUE TITLE
Improve waterfall readability

### ISSUE THREAD

**Issue by pmeenan** [WPO-Foundation/webpagetest#621]
From [forum post](http://www.webpagetest.org/forums/showthread.php?tid=14325):

```
I am often having a hard time locating the lines "Start Render", "msFirstPaint" and "DOM Interactive"
in the Waterfall View. Maybe because I am a bit color blind. I guess the readability could be
improved, maybe by doing one of these:

1) Use both color and line style – eg. make Start Render dashed, make msFirstPaint dotted
and keep DOM Interactive solid.

2) Include the legend in the image and connect the legend and actual lines by drawing a thin
line between them.

In some Waterfall Views I cannot find one of Start Render and msFirstPaint. Are these lines
sometimes rendered on top of each other? Maybe the dash pattern could match so both lines
are still visible?
```

Additionally, when there are user timing marks, if there are a lot of them it can clutter the waterfall.

**Comment by rviscomi** [comment]
Maybe a way to toggle the visibility of overlays like these lines? For example, use cases may be:

- show only user timing
- show only native events
- show only specific marks/events

This may be hard if the lines are rendered into the waterfall as an image. Any plans to HTMLify the waterfall?

**Comment by andydavies** [comment]
@rviscomi SVG seems the way to go wonder if we could adapt or integrate https://github.com/micmro/PerfCascade

There are also the complications of the waterfalls that are overlaid for compare, or have the scrolling red marker that follows the current filmstrip frame to account for too

**Comment by pmeenan** [comment]
I've considered SVG or drawing directly to canvas but the biggest issue is the copy/paste-ability of the waterfalls.  People copy/paste them a LOT and that becomes significantly more difficult even with SVG.  Maybe if I moved the "customize waterfall" link to above it and made it "view waterfall as image" but I'm still a bit worried that might be an additional step.

Right now I'm thinking something like this might work well:
* If there are > 2 user timing marks, don't show them in the image by default
* In the "customize waterfall" UI, allow for individual events to be turned on/off (including individual marks)
* When hovering over any of the metrics in the data table, draw a marker dynamically over the waterfall that shows the event (maybe make it sticky on clicking the data table element).  This would be pretty easy to do like with the moving red line in the filmstrip view.

**Comment by giovanisp** [comment]
Hi, out of curiosity, is this based on HAR Viewer (http://www.softwareishard.com/blog/har-viewer/) ? And/or could it be integrated ?
Il 12/ago/2016 14:27, Patrick Meenan <notifications@github.com> ha scritto:I've considered SVG or drawing directly to canvas but the biggest issue is the copy/paste-ability of the waterfalls.  People copy/paste them a LOT and that becomes significantly more difficult even with SVG.  Maybe if I moved the "customize waterfall" link to above it and made it "view waterfall as image" but I'm still a bit worried that might be an additional step.

Right now I'm thinking something like this might work well:

If there are > 2 user timing marks, don't show them in the image by defaultIn the "customize waterfall" UI, allow for individual events to be turned on/off (including individual marks)When hovering over any of the metrics in the data table, draw a marker dynamically over the waterfall that shows the event (maybe make it sticky on clicking the data table element).  This would be pretty easy to do like with the moving red line in the filmstrip view.

**Comment by pmeenan** [comment]
No, it isn't based on the HAR Viewer and there are a fair number of things that the HAR viewer can't do that the WPT UI needs because it is based on the original HAR spec which didn't allow for separation of DNS/Connect from the requests themselves (and it doesn't solve the copy/paste use case).

It would be possible to update the HAR viewer, integrate it, style it to match the WPT waterfalls and have the alternate "view as image" method but last time I looked at the code it was VERY involved since it was pulled out of Firebug.

**Comment by DebugXYZ** [comment]
I'm very interested in working on this, it really sounds like a fun challenge with unique requirements and this is one of those things that has been bugging me for quite a while. Once a basic implementation is put down with a bit of forethought, I can see some very useful (and fun to implement) possibilities when it comes to the comparison view.

Because of the amount / complexity of javascript. I'd need some guidance or confirmation when it comes to style and tooling choices before I give it a go.

For the initial version it would be an html table that is identical to the existing waterfall view. The table would only have two columns, name and waterfall. If we end up using a rendering framework, I'd likely implement sorting.

Things would then quickly become complicated when it comes to the waterfall view. I think the best choice would be canvas over svg due to browser quirks, and svg being very finicky from my experience. Also I feel much more comfortable working with canvas from a toy text editor I created a long while ago.

I like the idea of keeping it bare bones, but also don't want to neglect possibilities for the future. I am very unbiased when it comes to the js rendering framework and for the task at hand I can probably work with whatever is recommended, its only job would be to display a table and possibly sort/filter/add columns/ remove columns and maybe resize columns in the future.

If in the future we think we want more features, than it might be useful to start the good ol' js framework preference conversation. Adding a rendering framework also introduces a bit of a contribution hurdle and complexity but would be worth it if we think additional interactive features will be wanted.

Do we think we might want bells and whistles?

**Comment by DebugXYZ** [comment]
I think the framework choice could influence the coding style, but if not would something like using https://standardjs.com/ be acceptable?

**Comment by DebugXYZ** [comment]
I could also just do a mockup with html/css/canvas just to show where i would be going if that would be useful at first

**Comment by DebugXYZ** [comment]
Here is the prototype that is still in progress, any early feedback very welcome:

[https://debugxyz.github.io/wpt-wf-prototype/](https://debugxyz.github.io/wpt-wf-prototype/)

After feeling things out, I think a framework might actually be overboard.

I'm happy with the rough structure its self and a variation of it would likely end up in a PR, keeping each individual waterfall row as a separate canvas will simplify implementation of things like sorting, filtering, and comparison.

The main concept I want to bring is that, on the detail view (and other pages, like comparison, connection breakdown). The current image view would still be the default view. You would then need to click the "interactive view" link for more advanced features.

Some things I hope to add to the prototype:

* Sliders for scale/position
* Mousewheel for scaling and mouse dragging for position.
* Sorting
* Add the vertical timing markers
* generate table row structure from the wptRequestData object, to make things like sorting/filtering easier.
* comparison view - filter items that are in both waterfalls, filter items have significantly different timings (with a fuzz slider)

I'll likley leave cpu / bandwidth and main thread activity out of the prototype

Please let me know if the feature set / idea is horribly off and i'll know not to waist effort here.

**Comment by sburnicki** [comment]
I really appreciate your effort on working on the waterfall! I think it can bring up much potential for future features, just as those you described.

One of the most important properties of a new waterfall for me would be the request URL to be actual test, so we can use the browser search to look for a specific request.

What I'm a bit concerned about is that drawing a canvas for each row might bring up limitations in the future. If you think for example of a simple feature as drawing a vertical line in dashed style (see comments above), that may already look weird if a dashed line is drawn in each canvas separately. On the other hand, an "overlay" for overlapping drawings would bring up additional complexity and block other features as the selection individual rows.

In general I'm a fan of using SVGs as it brings features of both complex drawing possibilities and the advantages of using the DOM directly. I have positive experiences with using [d3.js](https://d3js.org/) for SVG drawing in other projects.
With SVG we'd be able to draw the whole waterfall as one graphic, without the need of dealing with both DOM manipulation and canvas drawing. But that's just my humble opinion ;)

Beside this I'd go for replicating the current waterfall features first, making a PR and adding new features aftwerwads (like sorting, properties, sliders, etc). This way we can get more early feedback and have several small PRs.

**Comment by soulgalore** [comment]
@DebugXYZ If you will go the SVG way I think you should join with @micmro and make https://github.com/micmro/PerfCascade more customizable so it could have the WebPageTest look and feel too.

**Comment by sburnicki** [comment]
About  @pmeenan's concern about copying and/or downloading a SVG waterfall:
We use [saveSvgAsPng](https://github.com/exupero/saveSvgAsPng) in OpenSpeedMonitor to export SVGs as PNG and it works quite good.
It basically converts the SVG to a standalone SVG (including all applied styles) and converts this into a data URI which is drawn on a canvas and converted to PNG by the browser.

So we could simply implement a download button or come up with some clever JS that creates the image and copies it to the clipboard directly.

**Comment by DebugXYZ** [comment]
I think d3.js could work out pretty well. I've worked with it lightly on a graphing project recently. I'd be happy to try out d3.js with a full drawing area and see how it fairs. I'm just trying to take cues from the project. Canvas kind of matches up with the existing implementation in its directness and think the maintainer would be more comfortable with this... but i'm just guessing here.

I looked at perfcascade and other HAR viewers before starting. I also spent quite a bit of time browsing through the existing webpagetest waterfall implementation. I'm concerned that webpagetest will have more unique requirements particularly when it comes to newer metrics that won't necessarily match up against the HAR format. There looks to be quite a bit in between the HAR format and webpagetest's waterfall.

I'm also concerned about having to go through two maintainers and have a feeling that requirements, purpose & directions won't line up for very simple things but again i'm just guessing here.

Some things that may be out of scope for most HAR viewers:

* Supporting a comparison view (the purpose of me wanting to spend some time on this)
* Filtering (toggling the vertical lines)
* Sorting
* adding additional columns
* Markers / timers that are not in the standard HAR format
* and probably a bunch of other interactive features
